pynamodb/attributes.py

```diff
@@ -807,25 +825,32 @@ def validate(self):
         return all(self.is_correctly_typed(k, v) for k, v in self.get_attributes().items())
 
     def serialize(self, values):
-        rval = {}
-        for k in values:
-            v = values[k]
-            if self._should_skip(v):
-                continue
-            attr_class = self._get_serialize_class(k, v)
-            if attr_class is None:
-                continue
-
-            # If this is a subclassed MapAttribute, there may be an alternate attr name
-            attr_name = attr_class.attr_name if not self.is_raw() else k
-
-            serialized = attr_class.serialize(v)
-            if self._should_skip(serialized):
-                # Check after we serialize in case the serialized value is null
-                continue
+        if isinstance(values, type(self)) and not values.is_raw():
+            return values._serialize()
 
-            rval[attr_name] = {attr_class.attr_type: serialized}
+        if not self.is_raw():
+            # Need to serialized based on the attributes in the class; however,
+            # the value passed in was not an instance of the class.
+            instance = type(self)()
+            instance.attribute_values = {}  # clear any defaults
+            for name in values:
+                if name in self.get_attributes():
+                    setattr(instance, name, values[name])
+            return instance._serialize()
```

Comment:
This fixes a bug where we did not check that required attributes were set since we iterated through the values to be serialized and not the attributes in the class. Now we "cast" whatever attribute container is passed in into the class we are serializing which performs the appropriate validation / null checks.